File: code/perl/search_ranges.pl
#!/usr/bin/env perl

use strict;
use warnings;

use JSON::MaybeXS;
use Search::Xapian ':all';
use FindBin qw($Bin);
use lib $Bin;
use Support;


my ($db_path, @terms) = @ARGV;
die "Usage: $0 DB_PATH QUERY..." unless $db_path && @terms;

search($db_path, join(' ', @terms));

sub search {
    my ($db_path, $query_string, $offset, $pagesize) = @_;
    $offset ||= 0;
    $pagesize ||= 10;
    my $db = Search::Xapian::Database->new($db_path);
    # Set up a QueryParser with a stemmer and suitable prefixes
    my $queryparser = Search::Xapian::QueryParser->new;
    $queryparser->set_stemmer(Search::Xapian::Stem->new('en'));
    $queryparser->set_stemming_strategy(STEM_SOME);

    # Start of prefix configuration.
    $queryparser->add_prefix(title => "S");
    $queryparser->add_prefix(description => "XD");
    # End of prefix configuration.

    # and add in range processors

    # this works with Search::Xapian on debian stable (1.2.24) and cpan (1.2.25)
    # for version in git master, method, classes and constant changed:
    # $queryparser->add_rangeprocessor(Xapian::NumberRangeProcessor->new(0, 'mm', RP_SUFFIX));
    # $queryparser->add_rangeprocessor(Xapian::NumberRangeProcessor->new(1);
    $queryparser->add_valuerangeprocessor(Search::Xapian::NumberValueRangeProcessor->new(0, 'mm', 0));
    $queryparser->add_valuerangeprocessor(Search::Xapian::NumberValueRangeProcessor->new(1));

    # And parse the query
    my $query = $queryparser->parse_query($query_string);

    # Use an Enquire object on the database to run the query
    my $enquire = $db->enquire($query);

    # And print out something about each match
    my @matches;

    my $mset = $enquire->get_mset($offset, $pagesize);
    foreach my $item ($mset->items) {
        my $fields = decode_json($item->get_document->get_data);
        printf(qq{%i: #%3.3i (%s) %s\n        %s\n},
               $item->get_rank + 1,
               $item->get_docid,
               $fields->{DATE_MADE},
               $fields->{MEASUREMENTS},
               $fields->{TITLE});
        push @matches, $item->get_docid;
    }
    Support::log_matches($query_string, $offset, $pagesize, \@matches);
}

File: code/c++/search_ranges.cc
#include <xapian.h>

#include <cstdio>
#include <iomanip>
#include <iostream>
#include <string>

#include "support.h"

using namespace std;

static void
search(const string & dbpath, const string & querystring,
       Xapian::doccount offset = 0, Xapian::doccount pagesize = 10)
{
    // offset - defines starting point within result set.
    // pagesize - defines number of records to retrieve.

    // Open the database we're going to search.
    Xapian::Database db(dbpath);

    // Set up a QueryParser with a stemmer and suitable prefixes.
    Xapian::QueryParser queryparser;
    queryparser.set_stemmer(Xapian::Stem("en"));
    queryparser.set_stemming_strategy(queryparser.STEM_SOME);
    queryparser.add_prefix("title", "S");
    queryparser.add_prefix("description", "XD");
    // and add in range processors
    Xapian::NumberRangeProcessor size_vrp(0, "mm", Xapian::RP_SUFFIX);
    queryparser.add_rangeprocessor(&size_vrp);
    Xapian::NumberRangeProcessor date_vrp(1);
    queryparser.add_rangeprocessor(&date_vrp);

    // And parse the query.
    Xapian::Query query = queryparser.parse_query(querystring);

    // Use an Enquire object on the database to run the query.
    Xapian::Enquire enquire(db);
    enquire.set_query(query);

    // And print out something about each match.
    Xapian::MSet mset = enquire.get_mset(offset, pagesize);

    clog << "'" << querystring << "'[" << offset << ":" << offset + pagesize
	 << "] =";
    for (Xapian::MSetIterator m = mset.begin(); m != mset.end(); ++m) {
	Xapian::docid did = *m;
	cout << m.get_rank() + 1 << ": #" << setfill('0') << setw(3) << did;

	const size_t DOC_FIELD_TITLE = 1;
	const size_t DOC_FIELD_DATE_MADE = 3;
	const size_t DOC_FIELD_MEASUREMENTS = 4;
	const string & data = m.get_document().get_data();
	cout << " (" << get_field(data, DOC_FIELD_DATE_MADE) << ") "
	     << get_field(data, DOC_FIELD_MEASUREMENTS) << "\n        "
	     << get_field(data, DOC_FIELD_TITLE) << '\n';
	// Log the document id.
	clog << ' ' << did;
    }
    clog << '\n';
}

int main(int argc, char** argv) {
    if (argc < 3) {
	cerr << "Usage: " << argv[0] << " DBPATH QUERYTERM...\n";
	return 1;
    }
    const char * dbpath = argv[1];

    // Join the rest of the arguments with spaces to make the query string.
    string querystring;
    for (argv += 2; *argv; ++argv) {
	if (!querystring.empty()) querystring += ' ';
	querystring += *argv;
    }

    search(dbpath, querystring);
}
